### PR TITLE
Update CI-tag-orchestrator for relay chain configs

### PR DESCRIPTION
**Pull request by yurikoinaba**
Build and publish a Docker image for both solo chain and relay chain configurations
- E2E tests run, but Docker images are published regardless of if they pass or fail (for now)
- Docker images are created for both solo chain configuration and relay chain configuration. No changes in tagging to solo chain configured images. For relay chain configured images, tags are appended with `-relay`, i.e. `0.6.0-relay`, `0.6.0-0.8.0-relay`, `latest-relay`
- Upload binaries as `zkv_runtime.solo.compact.compressed.wasm` and `zkv_runtime.relay.compact.compressed.wasm`

### CODE REVIEW

File: .github/workflows/CI-tag-orchestrator.yml [hunk 1]
```diff
@@ -120,7 +121,7 @@ jobs:
     strategy:
       matrix: 
         artifact_name:
-        - ${{ needs.build.outputs.artifact_name_solo }}
+        # - ${{ needs.build.outputs.artifact_name_solo }}
```
Comment (yurikoinaba) [inline .github/workflows/CI-tag-orchestrator.yml:124]: Will test w relay chain configured image first, and then run both

File: docker/scripts/bootstrap.sh [hunk 2]
```diff
@@ -25,18 +28,36 @@ if [ "$RELAY_FEATURES" ]; then
...
+  ${CARGO} build -p mainchain "${BUILD_PROFILE}"
 
-echo "----------------------------------------------------------"
-echo "Compile test parachain node"
-${CARGO} build -p paratest-node "${BUILD_PROFILE}"
+  echo "----------------------------------------------------------"
+  echo "Compile relay node"
+  ${CARGO} build -p zkv-relay "${BUILD_PROFILE}" "${RELAY_FEATURES}"
+
+  if [ "${SKIP_PARACHAIN}" != "true" ]; then
```
Comment (lander86) [inline docker/scripts/bootstrap.sh:55]: I think for the moment we will always skip parachain build.

File: docker/scripts/bootstrap.sh [hunk 3]
```diff
@@ -47,6 +68,8 @@ echo "----------------------------------------------------------"
 echo "Building relay node image"
 "${SCRIPTS}/build-zkv-relay-image-injected.sh"
 
-echo "----------------------------------------------------------"
-echo "Building parachain node image"
-"${SCRIPTS}/build-paratest-image-injected.sh"
+if [ "${SKIP_PARACHAIN}" != "true" ]; then
```
[the remark this answers is not in the document]
Comment (yurikoinaba) [reply docker/scripts/bootstrap.sh:71]: I wanted to keep it backwards compatible in case this was also used for testing other things

File: .github/workflows/CI-tag-orchestrator.yml [hunk 4]
```diff
@@ -5,54 +5,132 @@ on:
     tags:
       - "[0-9]+.[0-9]+.[0-9]+-[0-9]+.[0-9]+.[0-9]+*"
   workflow_dispatch:
+    inputs:
+      git_tag:
+        type: string
```
Comment (ysibirski) [inline .github/workflows/CI-tag-orchestrator.yml:10]: What is the reason to take user input for git tag? :thinking: setup_env.sh scrip is still checking if TAG exist in the upstream and derives from PROD or DEV branches.

File: .github/workflows/CI-tag-orchestrator.yml [hunk 5]
```diff
@@ -5,54 +5,132 @@ on:
...
+            target/
+          key: build-cache
       
-      - name: Build/Publish Docker image
-        env:
-          RELEASE_BRANCH: release
-          MAINTAINERS_KEYS: ${{ vars.MAINTAINERS_KEYS }}
-          DOCKER_HUB_USERNAME: ${{ secrets.DOCKER_HUB_USERNAME }}
-          DOCKER_HUB_TOKEN: ${{ secrets.DOCKER_HUB_TOKEN }}
+      - name: Verify release build
```
Comment (ysibirski) [inline .github/workflows/CI-tag-orchestrator.yml:63]:
```suggestion
      - name: Verify restored build cache artifacts
```

How about this instead if I understand this correctly

File: .github/workflows/CI-tag-orchestrator.yml [hunk 6]
```diff
@@ -5,54 +5,132 @@ on:
...
+          "${GITHUB_WORKSPACE}/docker/scripts/bootstrap.sh" --skip-parachain --skip-compile
+        shell: bash
+
+      - name: Set Docker image names as output
+        id: save-image-names
         run: |
-          source "${GITHUB_WORKSPACE}/ci/setup_env.sh"
-          "${GITHUB_WORKSPACE}/ci/docker.sh" --image-artifact zkverify-image
+          echo "docker_image_solo=horizenlabs/zkverify:latest" >> $GITHUB_OUTPUT
+          echo "docker_image_relay=horizenlabs/zkv-relay:latest" >> $GITHUB_OUTPUT
```
Comment (ysibirski) [inline .github/workflows/CI-tag-orchestrator.yml:79]: We should move `horizenlabs` into ENV vars better and then refer to this var under the steps

File: .github/workflows/CI-tag-orchestrator.yml [hunk 7]
```diff
@@ -5,54 +5,132 @@ on:
...
+          TAR_FILE_RELAY="${{github.workspace}}/${TAR_FILE_NAME_RELAY}.tar"
+
+          echo "tar_file_name_solo=$TAR_FILE_NAME_SOLO" >> $GITHUB_OUTPUT
+          echo "tar_file_solo=$TAR_FILE_SOLO" >> $GITHUB_OUTPUT
+
+          echo "tar_file_name_relay=$TAR_FILE_NAME_RELAY" >> $GITHUB_OUTPUT
+          echo "tar_file_relay=$TAR_FILE_RELAY" >> $GITHUB_OUTPUT
+
+          docker save "${{ steps.save-image-names.outputs.docker_image_solo }}" -o "${TAR_FILE_SOLO}"
+          docker save "${{ steps.save-image-names.outputs.docker_image_relay }}" -o "${TAR_FILE_RELAY}"
```
Comment (ysibirski) [inline .github/workflows/CI-tag-orchestrator.yml:97]: Why do we need `save-image-names` step if we can just use the values we are saving directly here:
`docker save horizenlabs/zkverify:latest -o ....` :thinking:

File: .github/workflows/CI-tag-orchestrator.yml [hunk 8]
```diff
@@ -5,54 +5,132 @@ on:
...
         run: |
-          source "${GITHUB_WORKSPACE}/ci/setup_env.sh"
-          "${GITHUB_WORKSPACE}/ci/docker.sh" --image-artifact zkverify-image
+          echo "docker_image_solo=horizenlabs/zkverify:latest" >> $GITHUB_OUTPUT
+          echo "docker_image_relay=horizenlabs/zkv-relay:latest" >> $GITHUB_OUTPUT
+
+      - name: Save the Docker image as a tarball
+        id: create-tar-file
+        run: |           
+          TAR_FILE_NAME_SOLO="zkverify-solo"
```
Comment (ysibirski) [inline .github/workflows/CI-tag-orchestrator.yml:84]: I think this can be an ENV var only for a particular scope therefore you do not need to save it under GITHUB_OUTPUT since it is static and not being dynamically defined during the `run:` block :shrug:

File: .github/workflows/CI-tag-orchestrator.yml [hunk 9]
```diff
@@ -5,54 +5,132 @@ on:
...
+          TAR_FILE_SOLO="${{github.workspace}}/${TAR_FILE_NAME_SOLO}.tar"
```
Comment (ysibirski) [inline .github/workflows/CI-tag-orchestrator.yml:85]: Same for this one

File: .github/workflows/CI-tag-orchestrator.yml [hunk 10]
```diff
@@ -5,54 +5,132 @@ on:
...
+
+          TAR_FILE_NAME_RELAY="zkverify-relay"
+          TAR_FILE_RELAY="${{github.workspace}}/${TAR_FILE_NAME_RELAY}.tar"
```
Comment (ysibirski) [inline .github/workflows/CI-tag-orchestrator.yml:88]: I think this can be an ENV var only for a particular scope therefore you do not need to save it under GITHUB_OUTPUT since it is static and not being dynamically defined during the run: block 🤷

File: .github/workflows/CI-tag-orchestrator.yml [hunk 11]
```diff
@@ -5,54 +5,132 @@ on:
...
+          echo "tar_file_relay=$TAR_FILE_RELAY" >> $GITHUB_OUTPUT
+
+          docker save "${{ steps.save-image-names.outputs.docker_image_solo }}" -o "${TAR_FILE_SOLO}"
+          docker save "${{ steps.save-image-names.outputs.docker_image_relay }}" -o "${TAR_FILE_RELAY}"
+        
+      - name: Upload Docker Image Artifact (solo chain)
+        uses: actions/upload-artifact@v4
+        with:
+          name: ${{ steps.create-tar-file.outputs.tar_file_name_solo }}
+          path: ${{ steps.create-tar-file.outputs.tar_file_solo }}
```
Comment (ysibirski) [inline .github/workflows/CI-tag-orchestrator.yml:103]: There is a chance these can be ENV VARS

File: .github/workflows/CI-tag-orchestrator.yml [hunk 12]
```diff
@@ -5,54 +5,132 @@ on:
...
+      CACHING_MODE: "CACHE_RESTORE"
+  
+  build:
     runs-on: ubuntu-latest
-    needs: execute-system-tests # Wait for e2e tests to finish
+    needs: env-setup
+
+    outputs:
+      artifact_name_solo: ${{ steps.create-tar-file.outputs.tar_file_name_solo }}
+      artifact_name_relay: ${{ steps.create-tar-file.outputs.tar_file_name_relay }}
```
Comment (ysibirski) [inline .github/workflows/CI-tag-orchestrator.yml:32]: May not be needed if using ENV vars for those values since they are static :shrug: Look at the comments below

File: .github/workflows/CI-tag-orchestrator.yml [hunk 13]
```diff
@@ -5,54 +5,132 @@ on:
...
+          overwrite: true
 
-      - name: Upload runtime artifact
+      - name: Upload Docker Image Artifact (relay chain)
         uses: actions/upload-artifact@v4
         with:
-          name: zkv_runtime.compact.compressed.wasm
-          path: ./zkv_runtime.compact.compressed.wasm
+          name: ${{ steps.create-tar-file.outputs.tar_file_name_relay }}
+          path: ${{ steps.create-tar-file.outputs.tar_file_relay }}
```
Comment (ysibirski) [inline .github/workflows/CI-tag-orchestrator.yml:113]: Same comment

File: .github/workflows/CI-tag-orchestrator.yml [hunk 14]
```diff
@@ -5,54 +5,132 @@ on:
...
+          compression-level: 0
+          overwrite: true
+
+  test-publish:
+    needs: build
+    strategy:
+      matrix: 
+        artifact_name:
+        - ${{ needs.build.outputs.artifact_name_solo }}
+        - ${{ needs.build.outputs.artifact_name_relay }}
```
Comment (ysibirski) [inline .github/workflows/CI-tag-orchestrator.yml:125]: Same here

File: .github/workflows/CI-tag-orchestrator.yml [hunk 15]
```diff
@@ -73,35 +151,35 @@ jobs:
...
-                      "text": "*Pull Request:* <${{ github.event.pull_request.html_url || github.event.head_commit.url }}|View PR/Commit>"
-                    }
-                  ]
-                }
-              ]
-            }
-        env:
-          SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
-          SLACK_WEBHOOK_TYPE: INCOMING_WEBHOOK
+  #     - name: Notify Slack
```
Comment (ysibirski) [inline .github/workflows/CI-tag-orchestrator.yml:154]: Still WIP for the commented out lines?

File: .github/workflows/CI-tag-orchestrator.yml [hunk 16]
```diff
@@ -5,54 +5,132 @@ on:
...
+    strategy:
+      matrix: 
+        artifact_name:
+        - ${{ needs.build.outputs.artifact_name_solo }}
+        - ${{ needs.build.outputs.artifact_name_relay }}
+    
+    uses: ./.github/workflows/CI-test-publish.yml
+    with:
+      artifact_name: ${{ matrix.artifact_name }}
+      git_tag: ${{ github.event.inputs.git_tag }}
```
Comment (ysibirski) [inline .github/workflows/CI-tag-orchestrator.yml:130]: What if this job was not triggered manually but rather by PUSH ON TAGS? Will this line be still happy :thinking: 
```
github.event.inputs.git_tag
```

File: .github/workflows/CI-test-publish.yml [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+      DOCKER_HUB_TOKEN:
+        required: true
+
+jobs:
+  execute-e2e-tests:
+    runs-on: ubuntu-latest
+    name: Execute E2E Tests
+    outputs:
+      docker_image_tag: ${{ steps.set-docker-tag.outputs.docker_image_tag }}
+      test_status: ${{ steps.runtest.outputs.test_status }}
```
Comment (ysibirski) [inline .github/workflows/CI-test-publish.yml:24]: Is it me or `runtest` has to be `run-test` = id on line #42?

File: .github/workflows/CI-test-publish.yml [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+        required: true
+      DOCKER_HUB_TOKEN:
+        required: true
+
+jobs:
+  execute-e2e-tests:
+    runs-on: ubuntu-latest
+    name: Execute E2E Tests
+    outputs:
+      docker_image_tag: ${{ steps.set-docker-tag.outputs.docker_image_tag }}
```
Comment (ysibirski) [inline .github/workflows/CI-test-publish.yml:23]: I do not see `set-docker-tag` step :thinking:

File: .github/workflows/CI-test-publish.yml [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+      - name: Download zkVerify Docker image artifact
+        uses: actions/download-artifact@v4
+        with:
+          name: ${{ inputs.artifact_name }}
+          path: ./
+
+      - name: Load Docker image
+        run: |
+          image_name="$(docker load -i "${GITHUB_WORKSPACE}/${image_artifact}.tar" | awk '/Loaded image:/ { print $3 }')"
+          log_info "=== Loaded image ${image_name} ==="
```
Comment (ysibirski) [inline .github/workflows/CI-test-publish.yml:38]: log_info is a function, where is it coming from :thinking:

File: .github/workflows/CI-test-publish.yml [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+
+      - name: Download zkVerify Docker image artifact
+        uses: actions/download-artifact@v4
+        with:
+          name: ${{ inputs.artifact_name }}
+          path: ./
+
+      - name: Load Docker image
+        run: |
+          image_name="$(docker load -i "${GITHUB_WORKSPACE}/${image_artifact}.tar" | awk '/Loaded image:/ { print $3 }')"
```
Comment (ysibirski) [inline .github/workflows/CI-test-publish.yml:37]: I do not see where `image_artifact` var is defined :thinking:

File: .github/workflows/CI-test-publish.yml [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+        with:
+          artifact_name: ${{ inputs.artifact_name }}
+          zkverify_docker_tag: latest
+          GH_TOKEN: ${{ secrets.GH_TOKEN }}
+
+  publish-docker-image:
+    name: Publish Docker Image
+    runs-on: ubuntu-latest
+    # needs: execute-e2e-tests # Wait for e2e tests to finish
+    # if: ${{ needs.execute-e2e-tests.outputs.test_status == 'success' }}
```
Comment (ysibirski) [inline .github/workflows/CI-test-publish.yml:53]: For future imlementation?

File: ci/docker.sh [hunk 22]
```diff
@@ -54,18 +53,15 @@ if [ "${is_a_release}" = "true" ]; then
...
+    log_info "=== Using Docker image artifact from upstream ==="
     image_name="$(docker load -i "${GITHUB_WORKSPACE}/${image_artifact}.tar" | awk '/Loaded image:/ { print $3 }')"
     log_info "=== Loaded image ${image_name} ==="
-    log_info "=== Renaming image ==="
-    docker tag "${image_name}" "${docker_hub_org}/${docker_image_build_name}:${docker_tag_full}"
   else 
-    log_info "=== Building Docker image: ${docker_hub_org}/${docker_image_build_name}:${docker_tag_full} ==="
-    docker build --build-arg PROFILE=production -f "${docker_file_path}" -t "${docker_hub_org}/${docker_image_build_name}:${docker_tag_full}" .
+    log_info "=== No artifact specified ==="
+    exit 1
```
Comment (ysibirski) [inline ci/docker.sh:64]: This should be `fn_die` vs `log_info` which will be in RED color and has exit 1 embedded. I would also add this smth about `--image-artifact` option for a better clarity of the error message

File: ci/docker.sh [hunk 23]
```diff
@@ -83,9 +79,16 @@ if [ -n "${docker_tag_full:-}" ]; then
     publish_tags=("${docker_tag_full}")
   fi
 
+  # Append -relay to tag names for relay chain images
+  if [[ "${image_artifact}" == *"relay"* ]]; then
```
Comment (ysibirski) [inline ci/docker.sh:83]: The idea is to catch ANYTHING that contains `relay` and append -relay to it? That means string like this will match as well `no_relay_here`. Just curious if we need smth more strict or not really?

File: ci/docker.sh [hunk 24]
```diff
@@ -83,9 +79,16 @@ if [ -n "${docker_tag_full:-}" ]; then
...
+    for publish_tag in "${!publish_tags[@]}"; do
+      publish_tags[$publish_tag]="${publish_tags[$publish_tag]}-relay"
```
Comment (ysibirski) [inline ci/docker.sh:85]:
```suggestion
      publish_tags["${publish_tag}"]="${publish_tags["${publish_tag}"]}-relay"
```

File: docker/scripts/bootstrap.sh [hunk 25]
```diff
@@ -25,18 +28,36 @@ if [ "$RELAY_FEATURES" ]; then
...
+        --skip-parachain)
+            SKIP_PARACHAIN=true
+            shift
+            ;;
+        --skip-compile)
+            SKIP_COMPILE=true
+            shift
+            ;;
+    esac
+done
```
Comment (ysibirski) [inline docker/scripts/bootstrap.sh:43]: If you want to fix the alignment 4 spaces vs 2

File: docker/scripts/my_cargo [hunk 26]
```diff
@@ -19,7 +19,7 @@ docker run --rm -ti \
     -v "${HOME}/.gitconfig:/.gitconfig" \
     -v /etc/passwd:/etc/passwd:ro \
     -v /etc/group:/etc/group:ro \
-    -v .:/usr/src/node \
+    -v $(pwd):/usr/src/node \
```
Comment (ysibirski) [inline docker/scripts/my_cargo:22]:
```suggestion
    -v "$(pwd)":/usr/src/node \
```

**Review by ysibirski — COMMENTED** [review]
See some comments

File: .github/workflows/CI-test-publish.yml [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+          RELEASE_BRANCH: release
+          MAINTAINERS_KEYS: ${{ vars.MAINTAINERS_KEYS }}
+          DOCKER_HUB_USERNAME: ${{ secrets.DOCKER_HUB_USERNAME }}
+          DOCKER_HUB_TOKEN: ${{ secrets.DOCKER_HUB_TOKEN }}
+        run: |
+          source "${GITHUB_WORKSPACE}/ci/setup_env.sh"
+          "${GITHUB_WORKSPACE}/ci/docker.sh" --image-artifact ${{ inputs.artifact_name }}
+        shell: bash
+
+      - name: Upload runtime artifact
```
Comment (DanieleDiBenedetto) [inline .github/workflows/CI-test-publish.yml:78]: Is this workflow called after the tag orchestrator one ? That's why you moved it here ?

File: .github/workflows/CI-test-publish.yml [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+        uses: actions/upload-artifact@v4
```
Comment (DanieleDiBenedetto) [inline .github/workflows/CI-test-publish.yml:79]: Moreover, if the same docker.sh is being called for both solo chain and relay chain image build, we might want to produce both runtime artifacts and differentiate them via a meaningful name (e.g. `zkv_runtime.solo.compact.compressed.wasm` for solo chain and `zkv_runtime.relay.compact.compressed.wasm` for relay-chain)

File: docker/scripts/bootstrap.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (DanieleDiBenedetto) [reply docker/scripts/bootstrap.sh:71]: Is this script used somewhere else in the workflows ? Otherwise I would keep it as it was before since it was meant to be used more for local testing

**Review by DanieleDiBenedetto — CHANGES_REQUESTED** [review]

File: .github/workflows/CI-tag-orchestrator.yml (hunk 4 above)
Comment (yurikoinaba) [reply .github/workflows/CI-tag-orchestrator.yml:10]: Yes, to allow for a manual trigger if needed.

It was mostly useful for testing, but I can also see a scenario where you want to rerun it. If a test failed and it gets fixed, for example.

File: .github/workflows/CI-tag-orchestrator.yml (hunk 12 above)
Comment (yurikoinaba) [reply .github/workflows/CI-tag-orchestrator.yml:32]: They need to be outputs to be able to be used in a different job 
https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/store-information-in-variables#passing-values-between-steps-and-jobs-in-a-workflow

File: .github/workflows/CI-test-publish.yml (hunk 28 above)
Comment (yurikoinaba) [reply .github/workflows/CI-test-publish.yml]: Yes, good idea.

I kept it here but I wanted to ask where these artifacts are used and it's needed.

File: .github/workflows/CI-test-publish.yml (hunk 27 above)
Comment (yurikoinaba) [reply .github/workflows/CI-test-publish.yml]: It's called from the CI-tag-orchestrator now.

There is a matrix that runs build, test, and publish each for solo and relay chain
https://github.com/zkVerify/zkVerify/pull/171/files#diff-f93eb3b00b22939661e1b3f2100a5684cdea4aeb8580b8814d0b2780025d6ef9R27-R35

File: ci/docker.sh (hunk 23 above)
Comment (yurikoinaba) [reply ci/docker.sh:82]: It can be an exact match on ""zkverify-relay"

File: .github/workflows/CI-tag-orchestrator.yml (hunk 16 above)
Comment (yurikoinaba) [reply .github/workflows/CI-tag-orchestrator.yml:36]: Good question, it seems to be fine. I triggered this via tag push and didn't get errors on this line

https://github.com/zkVerify/zkVerify/actions/runs/11800656546

File: .github/workflows/CI-tag-orchestrator.yml (hunk 12 above)
Comment (ysibirski) [reply .github/workflows/CI-tag-orchestrator.yml]: I see why now. Your gitt_tag was in fact "" empty but build-and-test job is defaulting to github_ref..... tag coming from the job if input is empty, smth to keep in mind if this will not step on the way. When you manually INPUT one tag but `build-and-test` job will use a diff tag to build the project

image

```
steps:
      - uses: actions/checkout@v4
        with:
          ref: ${{ github.event.inputs.git_tag || github.ref }}
          fetch-depth: 0
```

File: .github/workflows/CI-test-publish.yml (hunk 28 above)
Comment (yurikoinaba) [reply .github/workflows/CI-test-publish.yml]: Will update the naming to what you suggested

File: .github/workflows/CI-tag-orchestrator.yml [hunk 29]
```diff
@@ -5,103 +5,87 @@ on:
...
   RUST_BACKTRACE: 1
 
 jobs:
-  execute-system-tests:
-    name: Execute System Tests
-    uses: HorizenLabs/zkVerify-qa/.github/workflows/CI-e2e-tests.yml@main
+  env-setup:
+    uses: ./.github/workflows/CI-build.yml
+    with:
+      CACHING_MODE: "CACHE_SAVE"
```
Comment (yurikoinaba) [inline .github/workflows/CI-tag-orchestrator.yml:24]: It seems to me that the build files should be generated again for this workflow, thus setting it to "CACHE_SAVE"

Or rather, it seems that these folders can be used from the cache
```            
~/.cargo/bin/
~/.cargo/registry/index/
~/.cargo/registry/cache/
~/.cargo/git/db/
```
But the `/target` folder should be regenerated.

Lmk if that is correct @DanieleDiBenedetto @lander86

File: .github/workflows/CI-test-publish.yml (hunk 28 above)
Comment (DanieleDiBenedetto) [reply .github/workflows/CI-test-publish.yml]: > Yes, good idea.
> 
> I kept it here but I wanted to ask where these artifacts are used and it's needed.

Yes these are the wasm binaries we use when performing a runtime upgrade

File: .github/workflows/CI-tag-orchestrator.yml (hunk 29 above)
Comment (DanieleDiBenedetto) [reply .github/workflows/CI-tag-orchestrator.yml:24]: @la10736

File: .github/workflows/CI-tag-orchestrator.yml (hunk 15 above)
Comment (la10736) [reply .github/workflows/CI-tag-orchestrator.yml:60]: I prefer if you remove them instead of comment out: git stand here exactly for doing this job :smile:

File: .github/workflows/CI-tag-orchestrator.yml (hunk 29 above)
Comment (la10736) [reply .github/workflows/CI-tag-orchestrator.yml:24]: Anyway I noted that caching mode is required (I cannot understand why... maybe is a wrong requirement)... so you can use `""` 

We should add to backlog if is the case to remove this requirement

File: .github/workflows/CI-build-test-publish.yml [hunk 30]
```diff
@@ -0,0 +1,159 @@
...
+        with:
+          path: |
+            target/
+          key: build-cache
+      
+      - name: Verify restored build cache
+        run: |
+          ls target
+          ls target/release
+
```
Comment (la10736) [inline .github/workflows/CI-build-test-publish.yml:63]: I don't think you need them to build docker images: please rid them.

File: .gitignore [hunk 31]
```diff
@@ -2,6 +2,7 @@ target/
 **/*.rs.bk
 .idea
 *.iml
+.secrets
```
Comment (la10736) [inline .gitignore:5]: What is the `secrets` about? (just curiosity)

File: docker/dockerfiles/zkv-relay.Dockerfile [hunk 32]
```diff
@@ -36,7 +36,7 @@ ARG PROFILE="release"
 ARG FEATURES=""
 
 ENV BINARY="${BINARY}" \
-  RUN_USER=hl
+  RUN_USER=user
```
Comment (la10736) [inline docker/dockerfiles/zkv-relay.Dockerfile:39]: That's ok.... but, why did you change it? I spent lot of time to catch the best one :smile: 
`hl` stand for horizen labs

File: docker/scripts/entrypoint.sh [hunk 33]
```diff
@@ -54,6 +54,7 @@ get_arg_name_from_env_name() {
     arg_name="${arg_name//_/-}"
     arg_name="${arg_name,,}"
     arg_name=--"${arg_name}"
+    arg_name=$(echo "$arg_name" | sed -r 's/^---+/--/')
```
Comment (la10736) [inline docker/scripts/entrypoint.sh:57]: Please, can you explain here why you need this replacement? When `arg_name` already contains some dashes in front of it?

**Review by la10736 — CHANGES_REQUESTED** [review]
Great, just remove 
- Commented code
- Useless steps in the build docker stage
- Take care the note about use and save cache on tag checking

File: docker/dockerfiles/zkv-relay.Dockerfile (hunk 32 above)
Comment (lander86) [reply docker/dockerfiles/zkv-relay.Dockerfile:39]: zkVerify has been moved to zkVerify org and EDS is the one working on the development. I think we should get rid of everything "hl".

File: docker/scripts/entrypoint.sh (hunk 33 above)
Comment (lander86) [reply docker/scripts/entrypoint.sh:57]: With this sed command I'm not adding the double hyphens but I'm catching the cases in which we get 3 hyphens and replacing them with just two (or it will error). This happens because the script is not robust and, if the parameters value has an underscore more, it will append more hyphens than expected (this was happening on zkv-relay). 
The script will be more radically updated later one by completely removing the substitution logic to have something more robust.

File: .github/workflows/CI-tag-orchestrator.yml (hunk 16 above)
Comment (yurikoinaba) [reply .github/workflows/CI-tag-orchestrator.yml:36]: I think best if this logic is used when the input is defined

File: .github/workflows/CI-build-test-publish.yml [hunk 34]
```diff
@@ -0,0 +1,149 @@
...
+          name: ${{ needs.build.outputs.artifact_name }}
+          path: ./
+
+      - name: Load Docker image and rename
+        run: |
+          IMAGE_NAME="$(docker load -i "${GITHUB_WORKSPACE}/${{ needs.build.outputs.artifact_name }}.tar" | awk '/Loaded image:/ { print $3 }')"
+          echo $(docker images -q "${IMAGE_NAME}" > /dev/null 2>&1 && echo "Image exists..." || echo "Image missing...")
+          echo "Loaded image ${IMAGE_NAME}"
+          TEST_TAG="test"
+          echo "TEST_TAG=${TEST_TAG}" >> $GITHUB_ENV
```
Comment (ysibirski) [inline .github/workflows/CI-build-test-publish.yml:101]: Why not to use `test` explicitly without the VAR if we are not setting it dynamically or changing via ENV vars :thinking:

File: .github/workflows/CI-build-test-publish.yml [hunk 35]
```diff
@@ -0,0 +1,149 @@
...
+          DOCKER_HUB_TOKEN: ${{ secrets. DOCKER_HUB_TOKEN }}
+          GH_TOKEN: ${{ secrets.GH_TOKEN }}
+          QA_SLACK_WEBHOOK_URL: ${{ secrets.QA_SLACK_WEBHOOK_URL }}
+
+  publish-docker-image:
+    name: Publish Docker Image
+    runs-on: ubuntu-latest
+    needs: build
+    # needs: [ build, execute-e2e-tests ] # Wait for e2e tests to finish
+    # if: ${{ needs.execute-e2e-tests.outputs.test_status == 'success' }}
```
Comment (ysibirski) [inline .github/workflows/CI-build-test-publish.yml:120]: Still need this?

File: docker/scripts/entrypoint.sh (hunk 33 above)
Comment (ysibirski) [inline docker/scripts/entrypoint.sh:57]:
```suggestion
    arg_name="$(echo "$arg_name" | sed -r 's/^---+/--/')"
```

**Review by ysibirski — APPROVED** [review]
LGTM with couple small comments and one suggestion
